app/code/Magento/Catalog/Model/Product/Image/Process/Queue.php

```diff
@@ -0,0 +1,322 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Catalog\Model\Product\Image\Process;
+
+use Magento\Catalog\Api\ProductRepositoryInterface;
+use Magento\Catalog\Model\Product;
+use Magento\Catalog\Model\Product\Image\Process\CacheFactory;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Resize Queue
+ *
+ * Resize catalog images in parallel forks (if available)
+ */
+class Queue
+{
+    /**
+     * Default max amount of processes
+     */
+    const DEFAULT_MAX_PROCESSES_AMOUNT = 1;
+
+    /**
+     * Resize state identifier for "resized" state
+     */
+    const STATE_RESIZED = 1;
+
+    /**
+     * @var array
+     */
+    private $products = [];
+
+    /**
+     * @var int[]
+     */
+    private $processIds = [];
+
+    /**
+     * @var int[]
+     */
+    private $inProgress = [];
+
+    /**
+     * @var int
+     */
+    private $maxProcesses;
+
+    /**
+     * @var ProductRepositoryInterface
+     */
+    private $productRepository;
+
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @var OutputInterface
+     */
+    private $output;
+
+    /**
+     * @var CacheFactory
+     */
+    private $imageCacheFactory;
+
+    /**
+     * @var int
+     */
+    private $start = 0;
+
+    /**
+     * @var array
+     */
+    private $state = [];
+
+    /**
+     * @var int
+     */
+    private $lastJobStarted = 0;
+
+    /**
+     * @var Product\Image\Process\Cache
+     */
+    private $imageCache;
+
+    /**
+     * @param ProductRepositoryInterface $productRepository
+     * @param ResourceConnection $resourceConnection
+     * @param CacheFactory $imageCacheFactory
+     * @param OutputInterface $output
+     * @param int $maxProcesses
+     */
+    public function __construct(
+        ProductRepositoryInterface $productRepository,
+        ResourceConnection $resourceConnection,
+        CacheFactory $imageCacheFactory,
+        OutputInterface $output,
+        $maxProcesses = self::DEFAULT_MAX_PROCESSES_AMOUNT
+    ) {
+        $this->productRepository = $productRepository;
+        $this->resourceConnection = $resourceConnection;
+        $this->output = $output;
+        $this->imageCacheFactory = $imageCacheFactory;
+        $this->maxProcesses = $maxProcesses;
+        $this->imageCache = $this->imageCacheFactory->create();
```

Comment:
I would suggest moving this out of the class constructor to the method where it is used. It is not considered a good practice to execute anything else then argument validation and assignment in the constructor.